File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        
        int[] ans = new int[nums2.length];

        Stack<Integer> mono = new Stack<>();
        HashMap<Integer,Integer> index = new HashMap<>();
        for(int i=nums2.length-1; i>=0; i--){
            int n = nums2[i];
            index.put(n,i);
                while(!mono.isEmpty() &&n > mono.peek()){
                    mono.pop();
                }
                if(!mono.isEmpty()){
                    ans[i] = mono.peek();
                }
                else{
                    ans[i] = -1;
                }
                mono.push(n);
            
        }
        for(int x : ans) System.out.print(x + " ");

        for(int i=0; i<nums1.length; i++){
            nums1[i] = ans[index.get(nums1[i])];
        }
        return nums1;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        index = {}
        n= len(nums2)
       
        for i in range(n - 1, -1, -1):
            
          
            n = nums2[i]
            index[n] = i
            while stack and n > stack[-1]:
                stack.remove(stack[-1])
            if stack:
                nums2[i] = stack[-1]
            else:
                nums2[i]=-1
            stack.append(n)
           
        
        print(nums2)
        for i in range(0, len(nums1)):
            nums1[i] = nums2[index[nums1[i]]]

        return nums1
           
       
        
